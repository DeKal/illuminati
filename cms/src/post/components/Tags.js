import React from 'react'

import PropTypes from 'prop-types'
import Input from '@material-ui/core/Input'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import Chip from '@material-ui/core/Chip'
import { arrayToAttributes } from 'core/utils/misc'

const ITEM_HEIGHT = 48
const ITEM_PADDING_TOP = 8
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 250
    }
  }
}

const Tags = ({ isEdited, allTags, tags, editPostTags, classes }) => {
  return (
    <div className={classes.formControl}>
      <InputLabel>Tags</InputLabel>
      <Select
        multiple
        fullWidth
        value={tags}
        onChange={event => editPostTags(arrayToAttributes(event.target.value))}
        error={isEdited}
        input={<Input value={[]} />}
        renderValue={selected => renderSelectedChip(classes, selected)}
        MenuProps={MenuProps}
      >
        {allTags.map(tag => (
          <MenuItem key={tag} value={tag}>
            {tag}
          </MenuItem>
        ))}
      </Select>
    </div>
  )
}

export const renderSelectedChip = (classes, selected) => (
  <div className={classes.chips}>
    {selected.map(value => (
      <Chip key={value} label={value} className={classes.chip} />
    ))}
  </div>
)

export default Tags

Tags.propTypes = {
  isEdited: PropTypes.bool,
  allTags: PropTypes.array,
  tags: PropTypes.array,
  editPostTags: PropTypes.func,
  classes: PropTypes.shape({
    formControl: PropTypes.string,
    chip: PropTypes.string,
    chips: PropTypes.string
  })
}
