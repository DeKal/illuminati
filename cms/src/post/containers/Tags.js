import React from 'react'
import { connect } from 'react-redux'
import { makeStyles } from '@material-ui/core/styles'
import { getPostTags, getEditedTags } from 'post/state/selectors'
import { editPostTags } from 'post/state/actions'
import Tags from 'post/components/Tags'
import { attributesToArray } from 'core/utils/misc'
import isEqual from 'lodash-es/isEqual'

const useStyles = makeStyles(theme => ({
  formControl: {
    margin: theme.spacing(1)
  },
  chips: {
    display: 'flex',
    flexWrap: 'wrap'
  },
  chip: {
    margin: 2
  },
  noLabel: {
    marginTop: theme.spacing(3)
  }
}))

const TagsWithStyle = props => {
  const classes = useStyles()
  return <Tags {...props} classes={classes} />
}

const mapStateToProps = state => {
  const editedTags = getEditedTags(state)
  const tags = getPostTags(state)
  return {
    allTags: attributesToArray(tags) || [],
    tags: attributesToArray(editedTags) || [],
    isEdited: Object.keys(editedTags).length > 0 && !isEqual(editedTags, tags)
  }
}

const mapDispatchToProps = {
  editPostTags
}

export default connect(mapStateToProps, mapDispatchToProps)(TagsWithStyle)
