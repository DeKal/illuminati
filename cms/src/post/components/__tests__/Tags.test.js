import * as React from 'react'
import { shallow } from 'enzyme'
import Tags from '../Tags'
import { renderSelectedChip } from '../Tags'
const classes = {
  chip: 'chip',
  chips: 'chips',
  formControl: 'formControl'
}
const allTags = ['Algorithm', 'Spring MVC']

describe('Tags', () => {
  it('should render Tags correctly', function() {
    const wrap = shallow(<Tags classes={classes} allTags={allTags} />)
    expect(wrap).toMatchSnapshot()
  })

  it('should render tag chips correctly', function() {
    const wrap = shallow(renderSelectedChip(classes, allTags))
    expect(wrap).toMatchSnapshot()
  })
})
