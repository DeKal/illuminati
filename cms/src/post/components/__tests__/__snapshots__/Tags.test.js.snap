// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags should render Tags correctly 1`] = `
<div
  className="formControl"
>
  <WithStyles(ForwardRef(InputLabel))>
    Tags
  </WithStyles(ForwardRef(InputLabel))>
  <WithStyles(ForwardRef(Select))
    MenuProps={
      Object {
        "PaperProps": Object {
          "style": Object {
            "maxHeight": 224,
            "width": 250,
          },
        },
      }
    }
    fullWidth={true}
    input={
      <ForwardRef(WithStyles)
        value={Array []}
      />
    }
    multiple={true}
    onChange={[Function]}
    renderValue={[Function]}
  >
    <WithStyles(ForwardRef(MenuItem))
      key="Algorithm"
      value="Algorithm"
    >
      Algorithm
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="Spring MVC"
      value="Spring MVC"
    >
      Spring MVC
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Select))>
</div>
`;

exports[`Tags should render tag chips correctly 1`] = `
<div
  className="chips"
>
  <WithStyles(ForwardRef(Chip))
    className="chip"
    key="Algorithm"
    label="Algorithm"
  />
  <WithStyles(ForwardRef(Chip))
    className="chip"
    key="Spring MVC"
    label="Spring MVC"
  />
</div>
`;
